File: javascript/14.js
let lengths = [0,1,2]
let maxLen = 2;


for(let number=1000000; number>500000; number--){
    let localNumber = number;
    let len=0;
    
    while(localNumber>0){
        if(!lengths[localNumber]){
            if(localNumber%2==0){
                localNumber=localNumber/2;
            }else{
                localNumber=3*localNumber+1
            }
            len++;
        }else{
            len+=lengths[localNumber]
            lengths[number]=len;
            break
        }

    }
    
    maxLen = len > maxLen ? len : maxLen;    
}

console.table(lengths.indexOf(maxLen));

File: python/14.py
lengths = {2: 2}
maxLength = 2
maxNumber = 2

for i in range(1000000,500000,-1):
    _len = 0
    number = i

    while number > 0:

        if number in lengths:
            _len+=lengths[number]
            lengths[i]=_len
            break
        else:
            if number % 2 == 0:
                number /= 2
            else:
                number = 3*number + 1
            _len+=1

    if _len > maxLength:
        maxLength = _len
        maxNumber = i

print(maxLength,maxNumber)

